datafusion/core/src/physical_plan/join_utils.rs

```diff
@@ -296,6 +299,154 @@ impl<T> Clone for OnceFut<T> {
+                num_rows: cardinality,
+                // We don't do anything specific here, just combine the existing
+                // statistics which might yield subpar results (although it is
+                // true, esp regarding min/max). For a better estimation, we need
+                // filter selectivity analysis first.
+                column_statistics: all_left_col_stats
+                    .into_iter()
+                    .chain(all_right_col_stats.into_iter())
+                    .collect(),
+            })
+        }
+
+        JoinType::Semi => None,
+        JoinType::Anti => None,
+    }
+}
+
+/// Estimate the inner join cardinality by using the basic building blocks of
+/// column-level statistics and the total row count. This is a very naive and
+/// a very conservative implementation that can quickly give up if there is not
+/// enough input statistics.
+fn inner_join_cardinality(
+    left_num_rows: usize,
+    right_num_rows: usize,
+    left_col_stats: Vec<ColumnStatistics>,
+    right_col_stats: Vec<ColumnStatistics>,
+) -> Option<usize> {
+    // The algorithm here is partly based on the non-histogram selectivity estimation
+    // from Spark's Catalyst optimizer.
+
+    let mut join_selectivity = None;
+    for (left_stat, right_stat) in left_col_stats.iter().zip(right_col_stats.iter()) {
+        if (left_stat.min_value.clone()? > right_stat.max_value.clone()?)
+            || (left_stat.max_value.clone()? < right_stat.min_value.clone()?)
+        {
+            // If there is no overlap, then we can not accurately estimate
+            // the join cardinality. We could in theory use this information
+            // to point out the join will not produce any rows, but that would
+            // require some extra information (namely whether the statistics are
+            // exact). For now, we just give up.
+            return None;
+        }
+
+        let col_selectivity = max(left_stat.distinct_count, right_stat.distinct_count);
+        if col_selectivity > join_selectivity {
+            // Seems like there are a few implementations of this algorithm that implement
+            // exponential decay for the selectivity (like Hive's Optiq Optimizer). Needs
+            // further exploration.
+            join_selectivity = col_selectivity;
+        }
+    }
+
+    // With the assumption that the smaller input's domain is generally represented in the bigger
+    // input's domain, we can calculate the inner join's cardinality by taking the cartesian product
+    // of the two inputs and normalizing it by the selectivity factor.
+    let cardinality = match join_selectivity {
+        Some(selectivity) if selectivity > 0 => {
+            (left_num_rows * right_num_rows) / selectivity
+        }
+        // Since we don't have any information about the selectivity,
+        // we can only assume that the join will produce the cartesian
+        // product.
+        _ => left_num_rows * right_num_rows,
```

Comment:
For what it is worth, in practice joins where the cost matters are almost *never* cartesian products (because if they were for large inputs the query would never finish and if the inputs are small it doesn't really matter where in the join tree they go.

Reply:
We could also give up here (instead of trying to provide a conservative estimate), and see if we can come up with a more reasonable solution. The only case this happens is when we don't know the uniqueness of some join columns. @Dandandan and I have been discussing about an algorithm to actually use min/max values (and the total number of rows) to see if we can somehow infer distinct_count without having it in the statistics. That would hopefully make this case redundant. 

Filed #3813 for this.

Reply:
I think it might be better to give in that case?

There is also this presentation about optimizing the order of joins without statistics available (which also seems to do fine). We could also see if we can reuse some of these ideas:
 
https://www.youtube.com/watch?v=aNRoR0Z3SzU

Reply:
Makes sense! I've pushed https://github.com/apache/arrow-datafusion/pull/3787/commits/9e4467e76daab64432f5efeb8507f8761fe82dc0 which should turn this off now. We can work on #3813 to also improve this case when the distinct count is not known and min/max is known (which seems common) and then also start looking into the solution from the DuckDB. (thanks a lot for the link btw, it seems super interesting!)